sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessorEvent.cs

```diff
@@ -27,36 +29,89 @@ public struct EventProcessorEvent
         public EventData Data { get; }
 
         /// <summary>
-        ///   The function to call on checkpoint update.
+        ///   The <see cref="EventProcessorClient" /> for this instance to use as a way of checkpoiting.
         /// </summary>
         ///
-        private Func<EventData, PartitionContext, Task> OnUpdateCheckpoint { get; }
+        private WeakReference<EventProcessorClient> Processor { get; }
+
+        /// <summary>
+        ///   The callback to be called upon <see cref="UpdateCheckpointAsync" /> call.
+        /// </summary>
+        ///
+        private Func<EventData, PartitionContext, Task> UpdateCheckpointDelegate { get; }
 
         /// <summary>
         ///   Initializes a new instance of the <see cref="EventProcessorEvent"/> structure.
         /// </summary>
         ///
         /// <param name="partitionContext">The context of the Event Hub partition this instance is associated with.</param>
         /// <param name="eventData">The received event to be processed.  Expected to be <c>null</c> if the receive call has timed out.</param>
-        /// <param name="onUpdateCheckpoint">The function to call on checkpoint update.</param>
+        /// <param name="processor">The <see cref="EventProcessorClient" /> for this instance to use as a way of checkpoiting.</param>
+        ///
+        internal EventProcessorEvent(PartitionContext partitionContext,
+                                     EventData eventData,
+                                     EventProcessorClient processor)
+        {
+            Argument.AssertNotNull(partitionContext, nameof(partitionContext));
+            Argument.AssertNotNull(processor, nameof(processor));
+
+            Context = partitionContext;
+            Data = eventData;
+            Processor = new WeakReference<EventProcessorClient>(processor);
+            UpdateCheckpointDelegate = default;
+        }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="EventProcessorEvent"/> structure.
+        /// </summary>
+        ///
+        /// <param name="partitionContext">The context of the Event Hub partition this instance is associated with.</param>
+        /// <param name="eventData">The received event to be processed.  Expected to be <c>null</c> if the receive call has timed out.</param>
+        /// <param name="updateCheckpointImplementation">The callback to be called upon <see cref="UpdateCheckpointAsync" /> call.</param>
         ///
         public EventProcessorEvent(PartitionContext partitionContext,
```

Comment:
Thank you.   I still hate that we have to do this, but given how frequently we'll throw this type around,  I think the `struct` benefits outweigh the implementation warts.